src/compile/data/assemble.ts

```diff
@@ -0,0 +1,238 @@
+import {every, vals} from '../../util';
+import {VgData} from '../../vega.schema';
+import {AggregateNode} from './aggregate';
+import {BinNode} from './bin';
+import {DataFlowNode, OutputNode} from './dataflow';
+import {FacetNode} from './facet';
+import {ParseNode} from './formatparse';
+import {NonPositiveFilterNode} from './nonpositivefilter';
+import {NullFilterNode} from './nullfilter';
+import * as optimizers from './optimizers';
+import {optimizeFromLeaves} from './optimizers';
+import {OrderNode} from './pathorder';
+import {SourceNode} from './source';
+import {StackNode} from './stack';
+import {TimeUnitNode} from './timeunit';
+import {CalculateNode, FilterNode} from './transforms';
+
+/**
+ * Start optimization path from the root. Useful for removing nodes.
+ */
+function optimizeFromRoots(node: DataFlowNode) {
+  // remove empty non positive filter
+  if (node instanceof NonPositiveFilterNode && every(vals(node.filter), b => b === false)) {
+    node.remove();
+  }
+
+  // remove empty null filter nodes
+  if (node instanceof NullFilterNode && every(vals(node.aggregator), f => f === null)) {
+    node.remove();
+  }
+
+  // remove output nodes that are not required
+  if (node instanceof OutputNode && !node.required) {
+    node.remove();
+  }
+
+  node.children.forEach(optimizeFromRoots);
+}
+
+
+/**
+ * Return all leaf nodes.
+ */
+function getLeaves(roots: DataFlowNode[]) {
+  const leaves: DataFlowNode[] = [];
+  function append(node: DataFlowNode) {
+    if (node.numChildren() === 0) {
+      leaves.push(node);
+    } else {
+      node.children.forEach(append);
+    }
+  }
+
+  roots.forEach(append);
+  return leaves;
+}
+
+/**
+ * Print debug information for dataflow tree.
+ */
+function debug(node: DataFlowNode) {
```

Comment:
This seems useful, but currently is never called.  I wonder if we should always call it, but check if log.level is DEBUG?

In any case, replace console.log with log.debug